src/main/java/io/aiven/kafka/auth/audit/UserActivity.java

```diff
@@ -0,0 +1,31 @@
+/**
+ * Copyright (c) 2020 Aiven, Helsinki, Finland. https://aiven.io/
+ */
+
+package io.aiven.kafka.auth.audit;
+
+import java.io.Serializable;
+import java.time.ZonedDateTime;
+import java.util.ArrayList;
+import java.util.List;
+
+class UserActivity implements Serializable {
+
+    public final ZonedDateTime activeSince;
+
+    public final List<UserOperation> operations;
+
+    public UserActivity() {
+        this.activeSince = ZonedDateTime.now();
+        this.operations = new ArrayList<>();
+    }
+
+    public void addOperation(final UserOperation userOperation) {
+        operations.add(userOperation);
```

Comment:
We don't need to have the full list of operations, only the set of unique operations. In other words, we need `Set` instead of `List` (and of course `hashCode` and `equals` for `UserOperation`).
This should have been better specified, sorry.

Reply:
@willyborankin please note that `hashCode` and `equals` are needed for `UserOperation`

Reply:
both for `UserActivity` and `UserOperation` good catch